src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs

```diff
@@ -474,58 +482,86 @@ private static bool IsValidIPv6Host(string hostText)
                     // [::1] is the shortest valid IPv6 host
                     if (i < 4)
                     {
-                        return false;
+                        BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
+                    }
+                    else
+                    {
+                        // Tail call
+                        ValidateHostPort(hostText, i + 1);
+                        return;
                     }
-                    return IsValidHostPort(hostText, i + 1);
                 }
 
                 if (!IsHex(ch) && ch != ':' && ch != '.')
                 {
-                    return false;
+                    BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
                 }
             }
 
             // Must contain a ']'
-            return false;
+            BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static bool IsValidHostPort(string hostText, int offset)
+        private static void ValidateHostPort(string hostText, int offset)
         {
-            if (offset == hostText.Length)
+            // Skip bounds check for accessing the [offset] element
+            if ((uint)offset >= (uint)hostText.Length)
             {
-                return true;
+                return;
             }
 
-            if (hostText[offset] != ':' || hostText.Length == offset + 1)
+            var firstChar = hostText[offset];
+            offset++;
+            if (firstChar != ':' || (uint)offset >= (uint)hostText.Length)
             {
                 // Must have at least one number after the colon if present.
-                return false;
+                BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
             }
 
-            for (var i = offset + 1; i < hostText.Length; i++)
+            // This do+if check rather than for loop is to elimitate the bounds check, since
+            // the Jit doesn't currently pick up on it when starting at a variable offset
+            do
```

Comment:
I don't think transforming simple for loops to complicated do/while loops is worthwhile when we can just wait for the JIT to handle this better.

Reply:
Might be waiting a long time https://github.com/dotnet/coreclr/issues/15723#issuecomment-357152332
>  it's probably going to take a while, the rangecheck phase is a bit of a mess and it is difficult and risky to extend it to handle new patterns.

The issue here is its a range check per character; and when looking at the valid characters array that's another range check.